interfaces/builtin/pwm_test.go

```diff
@@ -0,0 +1,182 @@
+`, nil)
+	s.gadgetPwmSlotInfo = gadgetInfo.Slots["my-pin"]
+	s.gadgetPwmSlot = interfaces.NewConnectedSlot(s.gadgetPwmSlotInfo, nil, nil)
+	s.gadgetMissingNumberSlotInfo = gadgetInfo.Slots["missing-number"]
+	s.gadgetMissingNumberSlot = interfaces.NewConnectedSlot(s.gadgetMissingNumberSlotInfo, nil, nil)
+	s.gadgetBadNumberSlotInfo = gadgetInfo.Slots["bad-number"]
+	s.gadgetBadNumberSlot = interfaces.NewConnectedSlot(s.gadgetBadNumberSlotInfo, nil, nil)
+	s.gadgetMissingChipNumberSlotInfo = gadgetInfo.Slots["missing-chip-number"]
+	s.gadgetMissingChipNumberSlot = interfaces.NewConnectedSlot(s.gadgetMissingChipNumberSlotInfo, nil, nil)
+	s.gadgetBadChipNumberSlotInfo = gadgetInfo.Slots["bad-chip-number"]
+	s.gadgetBadChipNumberSlot = interfaces.NewConnectedSlot(s.gadgetBadChipNumberSlotInfo, nil, nil)
+	s.gadgetBadInterfaceSlotInfo = gadgetInfo.Slots["bad-interface-slot"]
+	s.gadgetBadInterfaceSlot = interfaces.NewConnectedSlot(s.gadgetBadInterfaceSlotInfo, nil, nil)
+	s.gadgetPlugInfo = gadgetInfo.Plugs["plug"]
+	s.gadgetPlug = interfaces.NewConnectedPlug(s.gadgetPlugInfo, nil, nil)
+	s.gadgetBadInterfacePlugInfo = gadgetInfo.Plugs["bad-interface-plug"]
+	s.gadgetBadInterfacePlug = interfaces.NewConnectedPlug(s.gadgetBadInterfacePlugInfo, nil, nil)
+
+	osInfo := snaptest.MockInfo(c, `
+name: my-core
+version: 0
+type: os
+slots:
+    my-pin:
+        interface: pwm
+        chip-number: 10
+        number: 7
+`, nil)
+	s.osPwmSlotInfo = osInfo.Slots["my-pin"]
+	s.osPwmSlot = interfaces.NewConnectedSlot(s.osPwmSlotInfo, nil, nil)
+}
+
+func (s *PwmInterfaceSuite) TestName(c *C) {
+	c.Assert(s.iface.Name(), Equals, "pwm")
+}
+
+func (s *PwmInterfaceSuite) TestSanitizeSlotGadgetSnap(c *C) {
+	// pwm slot on gadget accepeted
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.gadgetPwmSlotInfo), IsNil)
+
+	// slots without number attribute are rejected
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.gadgetMissingNumberSlotInfo), ErrorMatches,
+		"pwm slot must have a number attribute")
+
+	// slots with number attribute that isnt a number
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.gadgetBadNumberSlotInfo), ErrorMatches,
+		"pwm slot number attribute must be an int")
+
+	// slots without number attribute are rejected
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.gadgetMissingChipNumberSlotInfo), ErrorMatches,
+		"pwm slot must have a chip-number attribute")
+
+	// slots with number attribute that isnt a number
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.gadgetBadChipNumberSlotInfo), ErrorMatches,
+		"pwm slot chip-number attribute must be an int")
+}
+
+func (s *PwmInterfaceSuite) TestSanitizeSlotOsSnap(c *C) {
+	// pwm slot on OS accepted
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.osPwmSlotInfo), IsNil)
+}
+
+func (s *PwmInterfaceSuite) TestSanitizePlug(c *C) {
+	c.Assert(interfaces.BeforePreparePlug(s.iface, s.gadgetPlugInfo), IsNil)
+}
+
+func (s *PwmInterfaceSuite) TestSystemdConnectedSlot(c *C) {
```

Comment:
I think we actually need another test case here for the apparmor snippets:

```go
func (s *PwmInterfaceSuite) TestApparmorConnectedSlot(c *C) {
	spec := &apparmor.Specification{}
	err := spec.AddConnectedPlug(s.iface, s.gadgetPlug, s.gadgetPwmSlot)
	c.Assert(err, IsNil)
	c.Assert(spec.Snippets(), testutil.Contains, `/sys/class/pwm/pwmchip10/pwm100/* rwk`)
}
```

Reply:
Yes please - thanks for raising this @anonymouse64

Reply:
@anonymouse64 Thanks for hints! I tried to apply your suggestion but I always got the following:
```
FAIL: pwm_test.go:155: PwmInterfaceSuite.TestApparmorConnectedSlot

pwm_test.go:159:
    c.Assert(spec.Snippets(), testutil.Contains, "/sys/class/pwm/pwmchip10/pwm100/* rwk")
... container map[string][]string = map[string][]string{}
... elem string = "/sys/class/pwm/pwmchip10/pwm100/* rwk"
... container has items of type []string but expected element is a string

OOPS: 1203 passed, 1 FAILED
--- FAIL: Test (0.94s)
```
so how can I fix this as I'm not familiar with test code actually.